Reconciliation model refactoring in accordance with the AleRax article

StefanFlaumberg opened the pull request:
**This pull request depends on the ? PR**

Together they introduce the following changes:
1. Organize the code so that the `MultiModel` class defines general-purpose functions and the `UndatedDLMultiModel` and `UndatedDTLMultiModel` classes define only the functions dependent on modelling transfer events. Thus, the functions `computeLogLikelihood`, `updateCLVs` and `sampleOriginationSpecies` (renamed from `sampleSpeciesNode`) are now defined in the `MultiModel` class.
2. Make sure that cached likelihoods are cleared upon model parameter changes. Thus, the functions `setAlpha`, `setRates` and `setHighways` now all call `invalidateAllSpeciesNodes()` and `resetCache()`. As cached CLVs are not reused in the current version, the `invalidateAllSpeciesNodes` function does not have any effect now, but the code is organized to easily make it work in the future.
3. Make sure that the branch event probabilities get updated when and only when needed. The `recomputeSpeciesProbabilities` function now is called only within the `beforeComputeCLVs` function (renamed form `beforeComputeLogLikelihood`), which in turn is called only within the `updateCLVs` function. Indeed, we need to call `updateCLVs()` only within the `computeLogLikelihood` and `sampleReconciliations` functions, which are used when the tree or the model has been altered or just initialized, and we need to update the branch event probabilities only and every time before computing the CLVs.
4. Make the `--memory-savings` option work properly. The original `deallocateMemory` function used `_dtlclvs.clear()`, however, this method does not free the allocated memory. Now the function uses reinitialization with an empty vector, which I think should be solving the problem (I haven't checked it, though).
5. Fix `UndatedDLMultiModel`. The model has not been working at all after some recent commits, and its original implementation had a number of problems which are now all fixed:
5.1. Add the missing origination probability multiplier in the `getLikelihoodFactor` function. 
5.2. Add support for the rate gamma categories.
5.3. Add support for likelihood caching.
5.4. Add support for the `--memory-savings` option.
5.5. Resort to the iterative calculation of branch extinction probabilities (`_uE`). The original `solveSecondDegreePolynome` function used a wrong formula for quadratic equation roots, yet using the right formula sometimes results in `_uE = 0.0`, expectedly leading to an error. Thus, the iterative calculation method (fixed point optimization) is the only solution here.
6. Fix `UndatedDTLMultiModel`. The original implementation of the model had a number of problems which are now all fixed (fix #16):
6.1. Add transfer probability for each rate gamma category in the `WeightedHighway` structure.
6.2. Simplify implementation of transfer-extinction probabilities and clade transfer likelihoods. Now the former are stored along with branch extinction probabilities (as `_tE` and `_uE` respectively) and the latter are part of the `DTLCLV` structure (as `_tq`). This exactly reproduces the description provided in the AleRax paper.
6.3. Simplify implementation of the transfer constraints. Now the `_transferCandidateSpeciesNodes` vector stores potential transfer destination species branches for each source species branch. The vector is calculated by the `updateTransferCandidates` function called within the `recomputeSpeciesProbabilities` function. The new implementation is easier and faster than the original implementation.
7. Fix the fraction missing (`_fm`) usage in both `UndatedDLMultiModel` and `UndatedDTLMultiModel`. I think there was a misconception regarding this parameter both in the code and in the paper. The `_fm` parameter vector can be specified with the `--fraction-missing-file` option for terminal species branches (_i.e._ extant species) and is useful for analyses including incomplete transcriptome/genome assemblies, as `_fm[e]` stands for the  probability that the extant species `e` has a gene copy, but the copy is not included in the current dataset, _i.e._ `P(unobserved|EVENT_Leaf)`. This definition has two important implications. First, `_fm` cannot appear anywhere in the `computeProbability` function, as CLVs can be calculated only for the **observed** gene copies and their gene ancestors. Second, an unobserved on-leaf speciation event looks like a copy extinction, thus for an extant species branch `_uE[e] = P(EVENT_L) + P(unobserved ^ EVENT_Leaf) = P(EVENT_L) + P(unobserved|EVENT_Leaf)P(EVENT_Leaf) = _pL[e] + _fm[e]*_pS[e]`. Now both implications are respected in the code.
8. Fix origination probabilities calculation for the `--origination` option in both `UndatedDLMultiModel` and `UndatedDTLMultiModel` so that the sum of origination probabilities (`_OP`) over all allowed species branches equals 1.0.
9. Introduce DL events to both `UndatedDLMultiModel` and `UndatedDTLMultiModel` so that they can be turned off with the `--no-dl` option. For `UndatedDTLMultiModel` the implementation depends also on modelling TL events (using `--no-tl` allows for the faster direct probability calculation method).
10. Add inference of DL and TL events for the case when the clade is a gene leaf and the branch is a species leaf (only S events were modelled for this case before in both `UndatedDLMultiModel` and `UndatedDTLMultiModel`).
11. Fix `recCell.blLeft` and `recCell.blRight` specifications for S, D and T events (they were mixed up in both `UndatedDLMultiModel` and `UndatedDTLMultiModel`).
12. Make sure that the `ScaledValue` variables get scaled whenever needed and that random sampling is error-proof. Regarding the latter: we use a random sampling threshold `T = full_sum * Random::getProba()` and sample `s_i` if `(s_1 + ... + s_i) > T`. Should the `Random::getProba()` function return a value too close to 1.0, then due to auto-rounding (the implementation of which is system-dependent) we can get into the situation where `T = full_sum` and, thus, we sample `s_i` if `(s_1 + ... + s_i) > full_sum` -- a condition that is never met. Now, to avoid this situation the newly introduced `MultiModel::getRandom` function converts the threshold `T` into zero in case `T = full_sum`.
13. Fix specification of the scenario `_lastEventType` buffer. Now the `_lastEventType` buffer is set only before the both `backtrace()` calls of S and D events, before the `backtrace()` call of SL events and before the destination species `backtrace()` call of T and TL events. The idea is that DL and destination-loss TL events are virtual (not sampled) and that T events do not affect the source branch in any way (as their source is usually an unobserved extinct sister species branch), thus none of these events should change our estimate of singleton genes on a given species branch.
14. Get rid of the `rootCorrection` multiplier in the `computeLogLikelihood` function. The correction is not needed and can result in LL > 0 (try to reconcile a gene tree and a species tree of the same topology). Indeed, as mentioned in Benoit's comment to the correction, while computing the root clade likelihood, we integrate over all gene root positions using clade split frequencies of the root clade as weights of these root positions. Thus, the computed root clade likelihood is, by definition, the total probability over possible gene tree rootings. Similarly, we integrate over all species branches where the gene tree could have originated and sum the respective root clade likelihoods weighted by the origination probabilities of those branches. The computed reconciliation likelihood is also the total probability, now over possible gene tree originations and rootings. There is no need for any further normalization, and no such normalization is mentioned in the AleRax paper.
The observed difference in the estimated likelihoods mentioned in Benoit's comment is, instead, a problem of `GTBaseReconciliationModel` in GeneRax which doesn't use weights while summing over root positions (but this is a job for an unrelated future PR to GeneRax).

These changes do not introduce any new features, but render the reconciliation models to work in accordance with the original description provided in the AleRax paper.

StefanFlaumberg commented:
Hi @noahares,

I've sent the results of the PR testing on your email (hopefully, they've reached you well).

The changes in the PR are well thought-out, but as they affect many key aspects of reconciliation model implementation, having an expert review is crucial here. If there appear any questions regarding the changes made, I'll be happy to facilitate the review process by answering them. Please contact me here or via my email (stefanflaumberg@gmail.com).

If everything is fine, I suggest this version be tagged and released (to close #17 ) as a stable version.
I believe this PR fixes most of the program's existing bugs. My previous large PR is about making the code more organized and the program more user-friendly and is not a great improvement on its own without this PR. Thus, I think it might be reasonable to reuse the v1.3.0 tag for this version (can be done with `git tag -a -f v1.3.0 -m 'release v1.3.0'`).

Best wishes,
Stefan

noahares commented:
Thank you very much @StefanFlaumberg! I will have a look at the changes and your testing results as soon as possible.

StefanFlaumberg commented:
Hi @noahares,
I've added the aforementioned changes.

**What's new:**
- fixed bug with reporting the O reconciliation rate.
- fixed bug that disabled checking families by default (`--skip-family-filtering` value was misinterpreted). 
- fixed bug with family trimming that caused non-zero code exit if the second (`--trim-ratio`) trimming step received no valid families to start with. Also the step's trimming logs should now be a bit more easy to understand.
- added more compatibility checks for input parameters.
- renamed `--gene-tree-sample-frequency` to `--ccp-sample-frequency` (an option to set the subsampling step for input gene tree lists), because the original name can easily be conflated with the `--gene-tree-samples` option. I'm not particularly fond of the new name either, but at least it is different, and the option is not mentioned in wiki so the change is unlikely to cause any confusion.
- added support for `--no-dup` option disabling D and DL events. It is safe only when run under the DTL model with TL events allowed, thus other usage cases are explicitly restricted by compatibility checks.

Though I haven't done it myself, I suggest the version label be updated to v1.3.1 (as it's been quite a while since the last update), and this version could be tagged and released as a stable version.

StefanFlaumberg commented:
Added a fix for the failure occurring on using `--origination ROOT` together with `--prune-species-tree`.
In this case `--origination ROOT` and `--origination LCA` should behave identically, because the root of a pruned species tree is the LCA of all covered species. So now `--origination ROOT` uses the root of the species tree by default and the root of the pruned species tree if run with `--prune-species-tree` (in both cases the root is stored in `BaseReconciliationModel._prunedRoot`).

noahares commented on src/ale/UndatedDLMultiModel.hpp:
```diff
@@ -371,7 +369,7 @@ bool UndatedDLMultiModel<REAL>::computeProbability(
   // - SL event (only on an internal species branch)
   if (!isSpeciesLeaf) {
-    temp = _dlclvs[cid][fc] * (_uE[gc] * _PS[ec]);
+    temp = _dlclvs[cid][fc] * _uE[gc] * _PS[ec];
```
Is there a good reason for removing these parentheses everywhere? From what I remember they are pretty important for numerical reasons because C++ performs multiplications left-to-right and `_uE` and `_PX` (rates) values are orders of magnitude larger than the likelihood values. So it is preferable to first multiply these larger values for less precision loss.

StefanFlaumberg replied:
@noahares,
Thanks for bringing it up! No, in fact, I was just confused by not seeing any clear pattern of using parentheses in the upstream version, so I decided to use them only to separate multiplication of ScaledValues and multiplication of doubles into different groups (which has not much effect anyway).
I agree that clade probabilities `_uq` have lower magnitudes than extinction probabilities `_uE` and rates `_PX` and that it is preferable to use the large-first rule. Yet, again, I am not sure where to use it:
1. I guess the parentheses are not needed in the `recomputeSpeciesProbabilities` function, because extinction probabilities always include `_PL` as a summand, so we should generally consider `_uE` to be at the same scale as `_PX` and there can be no preferred order (and this agrees with the upstream version code).
2. Should we use extra parentheses in the expressions for S and D events, e.g. `leftsubclade_uq[f] * (rightsubclade_uq[g] * (_PX[e] * freq))`? I'm asking because in the upstream version it is always the two small values, the clade probabilities, that get multiplied first (e.g. see line 205 in the original DL model file). Is it because of the symmetry between the left and right clades (this doesn't seem to be a good explanation to me)?

StefanFlaumberg replied:
@noahares,
Sorry, I'm not sure I get this:
> Multiplying the right one by a small double-scale value first potentially decreases the accuracy

From what I understand:
As we do not use `scale()` anywhere in the multiplication chain, we can think of all the terms as of doubles (either they are doubles, like `_PX`, or we just work with the double-valued member of the ScaledValue class, like `_uE` or clade probas). We should acknowledge that clade probas (their double-valued members) can have very small values (close to `JS_SCALE_THRESHOLD`) and tend to do so, while `_uE` and `_PX` are generally much larger. It is known that when multiplying small and large values the best order of multiplication in terms of accuracy is the descending order; this I will call the "large-first" rule (not an expert here, read it on the Internet, but seems reasonable).
So in the expressions we are dealing with, the order should arguably be `clade_proba * (clade_proba * (_PX * freq))`, which corresponds to the `small * (small * (large * large))` order, and `clade_proba * (_uE * _PX)`, which corresponds to the `small * (large * large)` order. Somewhat contrary to your suggestions.

Unfortunately, things are not that simple in the upstream code:
> So currently in upstream: non-clade probability values get grouped with parentheses and clade probabilities go to the front of the expression.

Take a look in the upstream `UndatedDTLMultiModel.hpp` file. You are writing about, for example, line 685. Here it is true. But then we have line 698-699, transfer events, where a clade transfer probability first gets multiplied with a double and then with another clade probability. Clade transfer probabilities, for being averaged values of some clade probabilities, have the same low magnitude as clade probabilities, so why is the difference? Then in line 738, for highways, we get back to multiplying clade probabilities with each other first. I see no difference between the transfer and highway situations (clade probabilities have the same magnitude; one clade is transferred while the other stays), yet the multiplication orders differ. And which one is correct?

I can imagine the following 3 variants:
1. Multiply with larger values first everywhere. Then we have
for S and D events:
`temp = left_uq * (right_uq * (_PX * freq))`
for T and highway events:
`temp = staying_uq * (transferred_uq * (_PT * freq))`
for loss events:
`temp = surviving_uq * (_uE * _PX)`
This treats left and right clades a little bit differently in the expressions for S and D events, but should give the best accuracy as we follow the "large-first" rule.
2. Multiply with larger values first, but acknowledge symmetries between left and right clades. Then we have
for S and D events:
`temp = left_uq * right_uq * (_PX * freq)`
for T and highway events:
`temp = staying_uq * (transferred_uq * (_PT * freq))`
for loss events:
`temp = surviving_uq * (_uE * _PX)`
This is the closest to the upstream code. This way we might lose some accuracy for S and D events, but, contrary to variant 1, we treat left and right clades equivalently. Such equivalence cannot be implied for transfers, because there is always a staying clade and a transferred clade and one can always (though arbitrary) predefine a rule in which order to multiply them. 
3. Take the product of the smaller values first. Then we have
for S and D events:
`temp = left_uq * right_uq * (_PX * freq)`
for T and highway events:
`temp = staying_uq * transferred_uq * (_PT * freq)`
for loss events:
`temp = surviving_uq * (_uE * _PX)`
This is the closest to what you suggest. I do not see any reason to do it this way. The product of the first multiplication is at risk of losing many significant bits before they can be used in the second multiplication. Conversely, doing, for example, `(transferred_uq * (_PT * freq)` as in two other variants above first would result in losing fewer significant bits at the first step, thus giving a higher accuracy at the second step.

We might be wasting too much time on this, but exactly because these things matter for computations when using large datasets, they are worth being properly understood and dealt with. I am really not sure which of the 3 variants above is the right one, but I hope we can come to a conclusion. I am willing to finish it within a day.
